core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -552,6 +552,16 @@ public GeoPoint readGeoPoint() throws IOException {
         }
     }
 
+    public <T extends Writeable> T readOptionalWritable(Supplier<T> supplier) throws IOException {
+        if (readBoolean()) {
+            T writable = supplier.get();
+            writable.readFrom(this);
+            return writable;
```

Comment:
Shouldn't this be `return writable.readFrom(this);` ? Otherwise we return the ref retrieved from the supplier, which can be IngestStats.PROTO